File: infer/tests/codetoanalyze/java/pulse/taint/Streams.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package codetoanalyze.java.pulse;

import java.io.IOException;
import java.io.InputStream;

public class Streams {

  int bufferSize = 1024;

  void copyBad() throws IOException {
    InputStream tainted = (InputStream) InferTaint.inferSecretSource();
    InferTaint.inferSensitiveSink(read(tainted.toString().getBytes()));
  }

  void copyBadFN() throws IOException {
    InputStream tainted = (InputStream) InferTaint.inferSecretSource();
    InferTaint.inferSensitiveSink(read(tainted));
  }

  void copyBad1FN() throws IOException {
    InputStream tainted = (InputStream) InferTaint.inferSecretSource();
    byte[] data = new byte[24];
    tainted.read(data);
    InferTaint.inferSensitiveSink(data);
  }

  void systemArrayCopyBadFN() throws IOException {
    InputStream tainted = (InputStream) InferTaint.inferSecretSource();
    byte[] data = read(tainted.toString().getBytes());
    byte[] buffer = new byte[bufferSize];
    System.arraycopy(data, 0, buffer, 0, data.length);
    InferTaint.inferSensitiveSink(buffer);
  }

  byte[] read(InputStream is) throws IOException {
    return read(is.toString().getBytes());
  }

  byte[] read(byte[] data) {
    byte[] buffer = new byte[bufferSize];
    for (int i = 0; i < data.length; ++i) {
      buffer[i] = data[i];
    }
    return buffer;
  }
}

File: infer/tests/codetoanalyze/kotlin/pulse/taint/Streams.kt
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
package codetoanalyze.kotlin.pulse

import codetoanalyze.kotlin.pulse.InferTaint.inferSecretSource
import codetoanalyze.kotlin.pulse.InferTaint.inferSensitiveSink
import java.io.IOException
import java.io.InputStream

class Streams {

  var bufferSize = 1024

  @Throws(IOException::class)
  fun copyBad() {
    val tainted = inferSecretSource() as InputStream
    inferSensitiveSink(read(tainted.toString().toByteArray()))
  }

  @Throws(IOException::class)
  fun copyBadFN() {
    val tainted = inferSecretSource() as InputStream
    inferSensitiveSink(read(tainted))
  }

  @Throws(IOException::class)
  fun copyBad1FN() {
    val tainted = inferSecretSource() as InputStream
    val data = ByteArray(24)
    tainted.read(data)
    inferSensitiveSink(data)
  }

  @Throws(IOException::class)
  fun systemArrayCopyBadFN() {
    val tainted = inferSecretSource() as InputStream
    val data = read(tainted.toString().toByteArray())
    val buffer = ByteArray(bufferSize)
    System.arraycopy(data, 0, buffer, 0, data.size)
    inferSensitiveSink(buffer)
  }

  @Throws(IOException::class)
  fun read(`is`: InputStream): ByteArray {
    return read(`is`.toString().toByteArray())
  }

  fun read(data: ByteArray): ByteArray {
    val buffer = ByteArray(bufferSize)
    for (i in data.indices) {
      buffer[i] = data[i]
    }
    return buffer
  }
}
